feat: initial Miyoo Mini Flip compatibility; hardware detection, lid close events, lid close settings options

xXJSONDeruloXx opened the pull request:
Long time Onion Enjoyer, first time PR-er. 

I do not have a MM flip yet to test, but these additions are all in reference to the implementation work done in [Allium](https://github.com/goweiwen/Allium) to accommodate the flip's hardware and implement features around the lid events. 

Please let me know if this is at all helpful, and what needs to be changed to get this in a merge ready state. Anyone with MM Flip hardware any testing would be greatly appreciated.

Sichroteph commented:
@xXJSONDeruloXx  Thank you very much for this PR. I was about to roll up my sleeves and start working on it, then I thought I should check if someone like you had already done it. I will test these additions and get back to you as soon as possible.

Sichroteph commented:
@xXJSONDeruloXx I’m curious how you created MainUI 285 clean and MainUI 285 expert ?

I am testing this PR slowly because I struggle to find free time, but I am still making progress step by step.
At the moment, not much is working. Resolution changes seem to cause issues and, according to the keymon logs, the lid is not being detected.
I will tackle these points one by one.

xXJSONDeruloXx commented:
@Sichroteph thanks for the review, updated this to point to same bin

Sichroteph commented:
Yes please, that would be great. We’ll make sure to credit both of you.
Thanks a lot for the awesome work 🙏
We’re understaffed on our side, but the flame isn’t completely out yet 🙂

xXJSONDeruloXx commented:
@Dark98 I can not find your fork, but if you like I would be happy to implement your working version of lid detection and co author the commits

xXJSONDeruloXx commented:
ok have my flip mini now, was able to test this and with recent commit I am having lid close and open sleep wake, and tweaks options to change from sleep to shutdown does appear to be working too, so lid shut shuts down

xXJSONDeruloXx commented:
Demo with the 3 lid close settings in tweaks: 

https://youtube.com/shorts/rROmtfiIfsg?si=E39PpKHJTRNzrDmE

xXJSONDeruloXx commented:
@Dark98 can you link me to your fork?

xXJSONDeruloXx commented:
per discussions in the discord, here is a release off this branch, will encourage testing and report back or point users to this PR to gather feedback 

https://github.com/xXJSONDeruloXx/Onion/releases/tag/v4.5-dev-pr-1860

Aemiii91 commented on src/keymon/keymon.c:
```diff
@@ -436,8 +480,10 @@ int main(void)
+    printf_debug("Device detected: DEVICE_ID=%d (283=MM, 285=Flip, 354=Plus)", DEVICE_ID);
 
-    if (DEVICE_ID == MIYOO354) {
+    if (IS_MIYOO_PLUS_OR_FLIP()) {
```
The `HAS_AXP` convenience macro could be used here as well:
```suggestion
    if (HAS_AXP()) {
```

Aemiii91 commented on src/keymon/keymon.c:
```diff
@@ -625,8 +691,8 @@ int main(void)
                     case SELECT:
-                        if (DEVICE_ID == MIYOO354)
-                            break; // disable this shortcut for MMP
+                        if (DEVICE_ID == MIYOO285 || DEVICE_ID == MIYOO354)
```
```suggestion
                        if (IS_MIYOO_PLUS_OR_FLIP())
```

Aemiii91 commented on src/keymon/keymon.c:
```diff
@@ -663,8 +729,8 @@ int main(void)
                     case SELECT:
-                        if (DEVICE_ID == MIYOO354)
-                            break; // disable this shortcut for MMP
+                        if (DEVICE_ID == MIYOO285 || DEVICE_ID == MIYOO354)
```
```suggestion
                        if (IS_MIYOO_PLUS_OR_FLIP())
```

Aemiii91 commented on src/tweaks/menus.h:
```diff
@@ -102,7 +102,7 @@ void menu_datetime(void *_)
         network_loadState();
 
-        if (DEVICE_ID == MIYOO354 || network_state.ntp) {
+        if (DEVICE_ID == MIYOO285 || DEVICE_ID == MIYOO354 || network_state.ntp) {
```
```suggestion
        if (IS_MIYOO_PLUS_OR_FLIP() || network_state.ntp) {
```
or even better, create an equivalent `HAS_WIFI()` macro!

Aemiii91 commented on static/build/.tmp_update/runtime.sh:
```diff
@@ -640,8 +648,14 @@ check_hide_recents() {
+    if [ $DEVICE_ID -eq $MODEL_MMF ]; then
+        mainui_model=$MODEL_MMP
+    fi
+    
```
```suggestion
```

Aemiii91 commented on static/build/.tmp_update/runtime.sh:
```diff
@@ -640,8 +648,14 @@ check_hide_recents() {
+    
     mainui_mode=$([ -f $sysdir/config/.showExpert ] && echo "expert" || echo "clean")
-    mainui_srcname="MainUI-$DEVICE_ID-$mainui_mode"
+    mainui_srcname="MainUI-$mainui_model-$mainui_mode"
```
```suggestion
    mainui_srcname="MainUI-$DEVICE_ID-$mainui_mode"
```

Aemiii91 commented on static/build/.tmp_update/runtime.sh:
```diff
@@ -734,8 +748,10 @@ init_system() {
+    if [ $DEVICE_ID -eq $MODEL_MMF ] || [ $DEVICE_ID -eq $MODEL_MMP ]; then
+        if [ -f $sysdir/config/.lcdvolt ]; then
+            $sysdir/script/lcdvolt.sh 2> /dev/null
+        fi
```
```suggestion
    if [ $DEVICE_ID -eq $MODEL_MMP ] && [ -f $sysdir/config/.lcdvolt ]; then
        $sysdir/script/lcdvolt.sh 2> /dev/null
```

xXJSONDeruloXx commented:
Addressed all comments/change requests and tested on my device. also went ahead and added the tenlevels headphone detection slightly modified for deviceid rather than hal sensor detect. Also on sleep to shutdown if set I am having that be respected on lid close shutdown, I think that grants the most consistent behavior across devices. if sleep to shutdown is not set, then lid close sleeps indefinitely as before. LMK if any other changes needed

Aemiii91 commented on src/keymon/keymon.c:
```diff
@@ -281,9 +284,13 @@ void suspend_exec(int timeout)
+    int suspend_lid_state = read_lid_state(); // Store initial lid state
+    
+    // Use shorter poll timeout for lid detection on flip devices
+    int poll_timeout = (DEVICE_ID == MIYOO285) ? 500 : timeout;
```
@xXJSONDeruloXx what about this one?

xXJSONDeruloXx replied:
It does, if no auto sleep, then on lid close if in tweaks it's set to suspend it just stays suspended till you open lid or battery dies. If do nothing or shutdown, same concept

xXJSONDeruloXx replied:
Copying from discord for posterity 


better battery life but then when sleep is set to off, and tweaks setting set to sleep on lid close, upon lid open the screen would not come back on because no poll. That would be inconsistent with lid open behavior if sleep set to 5 min or whatever else, where on lid open screen comes back up, lid open is treated as power button tap. 

What’s the preference here? I’d prefer consistent lid open behavior regardless of sleep timeout setting but lmk if you feel differently

xXJSONDeruloXx commented:
Can this merge?

Rom0067 commented:
Hello,

I apologize for necroing this thread but I was wondering what the best way is to install the latest version of onion with the miyoo mini flip enhancements.  Im sorry if this is a dumb question.  I looked around on the github and I even tried the latest version but it did not have the options for detecting lid close events.  Thank you so much for Onion its amazing on my miyoo mini.  I literally bought the flip for OnionOS!

starquake commented:
> Hello,
> 
> I apologize for necroing this thread but I was wondering what the best way is to install the latest version of onion with the miyoo mini flip enhancements. Im sorry if this is a dumb question. I looked around on the github and I even tried the latest version but it did not have the options for detecting lid close events. Thank you so much for Onion its amazing on my miyoo mini. I literally bought the flip for OnionOS!

Build the branch this PR merges into: https://github.com/OnionUI/Onion/tree/v4.5-dev

I think the docker build instructions are the easiest: https://onionui.github.io/docs/dev/setup

Rom0067 commented:
Thank you for replying.  I am familiar with Docker and the docker build instructions do seem pretty straight forward.  If I may ask three questions for clarification:

1.  After building this release do I install it normally on my sd card?
2. Can I install this without messing with my current settings/configuration? 
3. Just to clarify I would need to build with  make with-toolchain CMD=dev to use the most current dev build?


I really appreciate the help. Sorry for all the questions I have never done anything like this before and I just got my flip configured the way I like.  Everything is great but having the option to close the lid to power offf would be amazing.

Bapf commented:
>     1. After building this release do I install it normally on my sd card?
yes
>     2. Can I install this without messing with my current settings/configuration?
90% yes
(if you get issues because of this, then the reason most likely will be the V4.5 code itself, not the mini flip additions)
>     3. Just to clarify I would need to build with  make with-toolchain CMD=dev to use the most current dev build?
Because the changes from this fork are scheduled for the next (4.5), not the recent beta-version (4.4), an there is no downloadable build for this version yet.

Rom0067 commented:
> > 3. Just to clarify I would need to build with  make with-toolchain CMD=dev to use the most current dev build?
> > ```
> 
> Because the changes from this fork are scheduled for the next (4.5), not the recent beta-version (4.4), an there is no downloadable build for this version yet.

Thank you for the information!  In this last line the answer is yes I need to use the CMD=dev because that will download the latest dev version?  Sorry for being dense... thanks again!

starquake commented:
@Rom0067 No you do not need to use CMD=dev. Unless you want debug logging. But you probably don't.

Rom0067 commented:
I really appreciate everyones help with this. Unfortunetly I was unable to get it to work.  Thankfully I backed up my sd card before I tried.  I was able to successfully build the project with docker and copied everything from the dist folder over to my sd.  However the options for closing the lid did not show up in settings and most of my settings were wiped.  

I ended up using https://github.com/xXJSONDeruloXx/Onion/releases/tag/v4.5-dev-pr-1860 this fork and the lid closing options are now available.  I have a few minor bugs here and there but over all it seems to be working well for me.  

Again thanks for the help but I think the build option might be beyond my scope and I will happily wait for an official update.
